examples/other_datasets/plot_financial_data.py

```diff
@@ -0,0 +1,247 @@
+    # Vectorize covariance matrices by removing redundant information.
+    length = len(x) // 2
+    first = range(0, length)
+    last = range(len(x) - length, len(x))
+    down_cadrans = x[np.ix_(last, last)]
+    if keep_diagonal:
+        down_cadrans = [down_cadrans[i, j] for i in first for j in first if i <= j]
+    else:
+        down_cadrans = [down_cadrans[i, j] for i in first for j in first if i < j]
+    first_cadrans = np.reshape(x[np.ix_(last, first)], (1, len(x)))
+    ret = np.append(first_cadrans, down_cadrans)
+    return ret
+
+
+class SlimVector(TransformerMixin, BaseEstimator):
+    def __init__(self, keep_diagonal):
+        self.keep_diagonal = keep_diagonal
+
+    def fit(self, X, y):
+        return self
+
+    def transform(self, X):
+        return np.array([slim(x, self.keep_diagonal) for x in X])
+
+
+class OptionalWhitening(TransformerMixin, BaseEstimator):
+    def __init__(self, process=True, n_components=4):
+        self.process = process
+        self.n_components = n_components
+
+    def fit(self, X, y):
+        return self
+
+    def transform(self, X):
+        if not self.process:
+            return X
+        return Whitening(dim_red={"n_components": 4}).fit_transform(X)
+
+
+# Classical pipeline: put together the transformers, and add at the end
+# the classical SVM
+pipe = make_pipeline(
+    ToEpochs(n=10),
+    XdawnCovariances(nfilter=1),
+    OptionalWhitening(process=True, n_components=4),
+    SlimVector(keep_diagonal=True),
+    SVC(),
+)
+
+# Optimize the pipeline:
+# let's save some time and run the optimization with the classical SVM
+gs = GridSearchCV(
+    pipe,
+    param_grid={
+        "toepochs__n": [10, 20],
+        "xdawncovariances__nfilter": [1, 2],
+        "optionalwhitening__process": [True, False],
+        "optionalwhitening__n_components": [2, 4],
+        "slimvector__keep_diagonal": [True, False],
+    },
+    scoring="balanced_accuracy",
+)
+
+
+##############################################################################
+# Balance dataset
+# ---------------
+#
+# Balance the data and display the "ERP" [3]_.
+
+# Let's balance the problem using NearMiss.
+# Note: at this stage `features` also contains the `index` column.
+# So `NearMiss` we choose the closest 200 non-fraud epochs to the 200 fraud-epochs
+# based also on this `index` column. This should be improved for real use cases.
+X, y = NearMiss().fit_resample(features_scaled, target.to_numpy())
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.2, random_state=42
+)
+
+labels, counts = np.unique(y_train, return_counts=True)
+print(f"Training set shape: {X_train.shape}, genuine: {counts[0]}, frauds: {counts[1]}")
+
+labels, counts = np.unique(y_test, return_counts=True)
+print(f"Testing set shape: {X_test.shape}, genuine: {counts[0]}, frauds: {counts[1]}")
+
+# Before fitting the GridSearchCV, let's display the "ERP"
+epochs = ToEpochs(n=10).transform(X_train)
```

Comment:
Doc build fails on this line:
```
Traceback (most recent call last):
  File "/github/workspace/examples/other_datasets/plot_financial_data.py", line 206, in <module>
    epochs = ToEpochs(n=10).transform(X_train)
  File "/usr/local/lib/python3.9/site-packages/sklearn/utils/_set_output.py", line 157, in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
  File "/github/workspace/examples/other_datasets/plot_financial_data.py", line 114, in transform
    all_epochs = np.array(all_epochs)
ValueError: could not broadcast input array from shape (3,0) into shape (3,)
```

Reply:
Ok. I see. ToEpochs is failing. It is expecting a dataframe but got a numpy array instead.